Implement trusted types integration with setTimeout and setInterval

lukewarlow opened the pull request:
#### e3c49184f6ef80a04d39e1a9929e6cbb63d274e7
<pre>
Implement trusted types integration with setTimeout and setInterval
<a href="https://bugs.webkit.org/show_bug.cgi?id=267692">https://bugs.webkit.org/show_bug.cgi?id=267692</a>

Reviewed by Anne van Kesteren and Darin Adler.

Make use of StringContext IDL adaptor for TrustedScript within the ScheduledAction IDL converter.

Includes fix InterfaceName attribute within StringContext IDL generator code.

Also handles exceptions in IDL string conversion code within trustedTypeCompliantString converter.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval-expected.txt:
* Source/WebCore/bindings/js/JSDOMConvertScheduledAction.h:
(WebCore::Converter&lt;IDLScheduledAction&gt;::convert):
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(JSValueToNative):
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestScheduledActionRealConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestScheduledAction.cpp: Added.
(WebCore::JSTestScheduledActionDOMConstructor::prototypeForStructure):
(WebCore::JSTestScheduledActionDOMConstructor::initializeProperties):
(WebCore::JSTestScheduledActionPrototype::finishCreation):
(WebCore::JSTestScheduledAction::JSTestScheduledAction):
(WebCore::JSTestScheduledAction::createPrototype):
(WebCore::JSTestScheduledAction::prototype):
(WebCore::JSTestScheduledAction::getConstructor):
(WebCore::JSTestScheduledAction::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestScheduledActionPrototypeFunction_methodBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::JSTestScheduledAction::subspaceForImpl):
(WebCore::JSTestScheduledAction::analyzeHeap):
(WebCore::JSTestScheduledActionOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestScheduledActionOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestScheduledAction::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestScheduledAction.h: Added.
(WebCore::JSTestScheduledAction::create):
(WebCore::JSTestScheduledAction::createStructure):
(WebCore::JSTestScheduledAction::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestScheduledAction.idl: Copied from Source/WebCore/bindings/js/JSDOMConvertScheduledAction.h.

Canonical link: <a href="https://commits.webkit.org/276934@main">https://commits.webkit.org/276934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5aa8d76cb78d4449705ced44e73bef712a891296

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43469 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22500 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45880 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46104 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39595 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45773 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19918 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35937 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44043 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19630 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37447 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16978 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18987 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38513 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1525 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41675 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47648 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42721 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19921 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37733 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41388 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10230 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20101 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19552 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7441,8 +7441,8 @@ sub JSValueToNative
-    if ($type->extendedAttributes->{StringContext}) {
-        my $interfaceName = $interface->type->name;
+    if ($type->extendedAttributes->{StringContext} || $type->name eq "ScheduledAction") {
+        my $interfaceName = $codeGenerator->GetVisibleInterfaceName($interface);
```
This change handles IDL interfaces with an "InterfaceName" attribute.

lukewarlow commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -7441,8 +7441,8 @@ sub JSValueToNative
     push(@conversionArguments, $exceptionThrower) if $exceptionThrower;
-    if ($type->extendedAttributes->{StringContext}) {
-        my $interfaceName = $interface->type->name;
+    if ($type->extendedAttributes->{StringContext} || $type->name eq "ScheduledAction") {
```
This is needed so that we correctly pass the sink context through to ScheduledAction IDL converter.

darinadler replied:
We have a goal of not further hardcoding any type names into the bindings script. Is there a way to achieve this with an attribute instead of a known name?

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -154,6 +154,7 @@ String trustedTypeCompliantString(TrustedType expectedType, JSGlobalObject& glob
     auto stringValue = expectedType == TrustedType::TrustedScriptURL
         ? Converter<IDLUSVString>::convert(global, input)
         : Converter<IDLDOMString>::convert(global, input);
+    RETURN_IF_EXCEPTION(throwScope, { });
```
This line was needed else it triggered the a [test failure](https://ews-build.s3-us-west-2.amazonaws.com/macOS-AppleSilicon-Sonoma-Debug-WK2-Tests-EWS/abbfbb81-14986/js/dom/missing-exception-check-in-convertVariadicArguments-crash-log.txt)

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval.html:
```diff
@@ -19,7 +19,14 @@
-  trustedTypes.createPolicy("default", {createScript: s => "window." + s + ".done()"});
+  trustedTypes.createPolicy("default", {createScript: (s, _, sink) => {
+    if (s === "timeoutStringTest") {
+      assert_equals(sink, 'Window setTimeout');
```
This change adds missing test coverage to ensure the sink context is passed through correctly. (Previously it wasn't in this PR)

Export at https://github.com/web-platform-tests/wpt/pull/45058

lukewarlow committed:
Implement trusted types integration with setTimeout and setInterval
https://bugs.webkit.org/show_bug.cgi?id=267692

Reviewed by Anne van Kesteren and Darin Adler.

Make use of StringContext IDL adaptor for TrustedScript within the ScheduledAction IDL converter.

Includes fix InterfaceName attribute within StringContext IDL generator code.

Also handles exceptions in IDL string conversion code within trustedTypeCompliantString converter.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval-expected.txt:
* Source/WebCore/bindings/js/JSDOMConvertScheduledAction.h:
(WebCore::Converter<IDLScheduledAction>::convert):
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(JSValueToNative):
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestScheduledActionRealConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestScheduledAction.cpp: Added.
(WebCore::JSTestScheduledActionDOMConstructor::prototypeForStructure):
(WebCore::JSTestScheduledActionDOMConstructor::initializeProperties):
(WebCore::JSTestScheduledActionPrototype::finishCreation):
(WebCore::JSTestScheduledAction::JSTestScheduledAction):
(WebCore::JSTestScheduledAction::createPrototype):
(WebCore::JSTestScheduledAction::prototype):
(WebCore::JSTestScheduledAction::getConstructor):
(WebCore::JSTestScheduledAction::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestScheduledActionPrototypeFunction_methodBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::JSTestScheduledAction::subspaceForImpl):
(WebCore::JSTestScheduledAction::analyzeHeap):
(WebCore::JSTestScheduledActionOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestScheduledActionOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestScheduledAction::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestScheduledAction.h: Added.
(WebCore::JSTestScheduledAction::create):
(WebCore::JSTestScheduledAction::createStructure):
(WebCore::JSTestScheduledAction::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestScheduledAction.idl: Copied from Source/WebCore/bindings/js/JSDOMConvertScheduledAction.h.

Canonical link: https://commits.webkit.org/276934@main
